hazelcast/src/main/java/com/hazelcast/datalink/impl/DataLinkConsistencyChecker.java

```diff
@@ -0,0 +1,94 @@
+import com.hazelcast.jet.impl.JetServiceBackend;
+import com.hazelcast.map.IMap;
+import com.hazelcast.spi.impl.NodeEngine;
+import com.hazelcast.sql.impl.QueryUtils;
+import com.hazelcast.sql.impl.schema.datalink.DataLink;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class DataLinkConsistencyChecker {
+    private final DataLinkServiceImpl dataLinkService;
+    private final IMap<Object, Object> sqlCatalog;
+
+    public DataLinkConsistencyChecker(HazelcastInstance instance, NodeEngine nodeEngine) {
+        assert nodeEngine.getDataLinkService() instanceof DataLinkServiceImpl;
+        this.dataLinkService = (DataLinkServiceImpl) nodeEngine.getDataLinkService();
+        this.sqlCatalog = instance.getMap(JetServiceBackend.SQL_CATALOG_MAP_NAME);
+    }
+
+    /**
+     * Removes & alters all outdated data links in __sql.catalog
+     * and adds all missed data links to __sql.catalog
+     */
+    public void check() {
+        List<Map.Entry<String, DataLinkSourcePair>> sqlEntries = dataLinkService.getDataLinks()
+                .entrySet()
+                .stream()
+                .filter(en -> en.getValue().source == DataLinkSource.SQL)
+                .collect(Collectors.toList());
+
+        for (Map.Entry<String, DataLinkSourcePair> entry : sqlEntries) {
+            Object catalogValue = sqlCatalog.get(QueryUtils.wrapDataLinkKey(entry.getKey()));
+
+            // Data link is absent in sql catalog -> add it.
+            if (catalogValue == null) {
+                addEntryToCatalog(entry);
+            } else {
+                // Try to alter outdated data links
+                assert catalogValue instanceof DataLink;
+                DataLink catalogDataLink = (DataLink) catalogValue;
+                if (!dataLinksAreEqual(entry.getValue(), catalogDataLink)) {
+                    sqlCatalog.remove(QueryUtils.wrapDataLinkKey(entry.getKey()));
+                    addEntryToCatalog(entry);
+                }
+            }
+        }
+
+        // Data link is not present in service, but still present in sql.catalog
+        sqlCatalog.removeAll(e -> e.getValue() instanceof DataLink
+                && !dataLinkService.existsDataLink(((DataLink) e.getValue()).getName()));
+    }
+
+    // package-private for testing.
+    void addEntryToCatalog(Map.Entry<String, DataLinkSourcePair> entry) {
+        DataLinkSourcePair pair = entry.getValue();
+        DataLink dataLink = new DataLink(
+                pair.instance.getName(),
+                pair.instance.getConfig().getClassName(),
+                pair.instance.options());
+        sqlCatalog.put(QueryUtils.wrapDataLinkKey(entry.getKey()), dataLink);
+    }
+
+    private boolean dataLinksAreEqual(DataLinkSourcePair pair, DataLink catalogDataLink) {
+        DataLinkConfig catalogDLConfig = dataLinkService.toConfig(
+                catalogDataLink.getName(),
```

Comment:
`DataLink` has `getConfig()` method. Why not use it? Current implementation also ignores `shared` flag.

Reply:
> DataLink has getConfig() method.

`com.hazelcast.datalink.DataLink` has it, but `com.hazelcast.sql.impl.schema.datalink.DataLink` doesn't. 
Yes, naming is a problem, but our current conventions for schema entities clashes with non-schema data link name.